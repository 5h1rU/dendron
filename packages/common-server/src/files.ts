import {
  DNodeRaw,
  DNodeRawOpts,
  IDNode,
  Note,
  NoteData,
  NoteRawProps,
  genUUID
} from "@dendron/common-all";
import fs, { Dirent } from "fs";

import _ from "lodash";
import matter from "gray-matter";
import minimatch from "minimatch";
import path from "path";

interface FileMeta {
  name: string;
}

type getAllFilesOpts = {
  root: string;
  exclude?: string[];
};

export function fileMeta2Node(body: string, meta: FileMeta): Note {
  const title = meta.name;
  // read id from file or generate one based on thte tile
  const id = genUUID();
  const note = new Note({
    id,
    title,
    desc: "TODO",
    data: {
      schemaId: "-1"
    },
    body,
    fname: meta.name
  });
  return note;
}

export function deleteFile(fpath: string) {
  return fs.unlinkSync(fpath);
}

export function globMatch(patterns: string[] | string, fname: string): boolean {
  if (_.isString(patterns)) {
    return minimatch(fname, patterns);
  }
  return _.some(patterns, pattern => minimatch(fname, pattern));
}

export function getAllFiles(opts: getAllFilesOpts): Dirent[] | string[] {
  const { root } = _.defaults(opts, {
    exclude: [".git", "Icon\r", ".*"]
  });
  const allFiles = fs.readdirSync(root, { withFileTypes: true });
  return _.reject(
    allFiles.map(dirent => {
      const { name: fname } = dirent;
      // match exclusions
      if (
        _.some([dirent.isDirectory(), globMatch(opts.exclude || [], fname)])
      ) {
        return null;
      }
      return dirent.name;
    }),
    _.isNull
  ) as Dirent[] | string[];
}

export function mdFile2NodeProps(fpath: string): NoteRawProps {
  // NOTE: gray matter cache old date, need to pass empty options
  // to bypass
  // see https://github.com/jonschlinkert/gray-matter/issues/43
  const { data, content: body } = (matter.read(fpath, {}) as unknown) as {
    data: DNodeRawOpts<NoteData>;
    content: string;
  };
  const { name: fname } = path.parse(fpath);
  const dataProps = DNodeRaw.createProps({ ...data, fname, body });
  // DEBUG: data: {data}, fpath: {fpath}, dataProps: {dataProps}
  return dataProps;
}

// export function mdFile2Node(fpath: string): Note {
//   // NOTE: gray matter cache old date, need to pass empty options
//   // to bypass
//   // see https://github.com/jonschlinkert/gray-matter/issues/43
//   const { data, content: body } = (matter.read(fpath, {}) as unknown) as {
//     data: DNodeRawOpts;
//     content: string;
//   };
//   const { name } = path.parse(fpath);
//   if (!data.title) {
//     data.title = name.split(".").slice(-1)[0];
//   }
//   if (!data.id) {
//     data.id = genUUID();
//   }
//   const note = new Note({ ...data, body, fname: name });
//   return note;
// }

export function node2MdFile(node: IDNode, opts: { root: string }) {
  const { root } = opts;
  const { body, path: nodePath } = node;
  const meta = _.pick(node, [
    "id",
    "title",
    "desc",
    "updated",
    "created",
    "url",
    "path"
  ]);
  const parent = node.parent?.id || null;
  const children = node.children.map(c => c.id);
  const filePath = path.join(root, `${nodePath}.md`);
  return fs.writeFileSync(
    filePath,
    matter.stringify(body || "", { ...meta, parent, children })
  );
}
