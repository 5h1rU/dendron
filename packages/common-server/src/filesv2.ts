import {
  DNodeUtilsV2,
  NotePropsV2,
  NoteUtilsV2,
  SchemaModuleOptsV2,
  SchemaUtilsV2,
} from "@dendronhq/common-all";
import fs from "fs-extra";
import matter from "gray-matter";
import YAML from "js-yaml";
import path from "path";

export function file2Note(fpath: string): NotePropsV2 {
  const options: any = {
    engines: {
      yaml: {
        parse: (s: string) => YAML.safeLoad(s, { schema: YAML.JSON_SCHEMA }),
        stringify: (s: string) =>
          YAML.safeDump(s, { schema: YAML.JSON_SCHEMA }),
      },
    },
  };
  const { data, content: body } = matter(
    fs.readFileSync(fpath, { encoding: "utf8" }),
    options
  );
  const { name: fname } = path.parse(fpath);
  return DNodeUtilsV2.create({ ...data, fname, body, type: "note" });
}

export function note2File(note: NotePropsV2, vaultPath: string) {
  const { fname } = note;
  const ext = ".md";
  return fs.writeFile(
    path.join(vaultPath, fname + ext),
    NoteUtilsV2.serialize(note)
  );
}

export function schemaModule2File(
  schemaFile: SchemaModuleOptsV2,
  vaultPath: string,
  fname: string
) {
  const ext = ".schema.yml";
  return fs.writeFile(
    path.join(vaultPath, fname + ext),
    SchemaUtilsV2.serializeModule(schemaFile)
  );
}
