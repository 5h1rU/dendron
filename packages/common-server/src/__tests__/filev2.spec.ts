import {
  DNodeUtilsV2 as _du,
  SchemaUtilsV2 as _su,
} from "@dendronhq/common-all";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { tmpDir } from "../files";
import { file2Schema, schemaModuleProps2File } from "../filesv2";

describe("schemaModuleProps2File", () => {
  let root: string;

  beforeEach(() => {
    root = tmpDir().name;
  });

  it("root", async () => {
    const fname = "root";
    const sm = _su.createRootModuleProps(fname);
    await schemaModuleProps2File(sm, root, fname);
    const payload = fs.readFileSync(path.join(root, `${fname}.schema.yml`), {
      encoding: "utf8",
    });
    expect(payload).toMatchSnapshot();
  });

  it("non-root", async () => {
    const fname = "bond";
    const smp = _su.createModuleProps({ fname });
    const rootNote = smp.schemas["root"];
    const ch1 = _su.create({ id: "ch1", fname });
    smp.schemas["ch1"] = ch1;
    _du.addChild(rootNote, ch1);

    await schemaModuleProps2File(smp, root, fname);
    const payload = fs.readFileSync(path.join(root, `${fname}.schema.yml`), {
      encoding: "utf8",
    });
    expect(payload).toMatchSnapshot();
  });
});

describe("file2Schema", () => {
  let root: string;
  beforeEach(() => {
    root = tmpDir().name;
  });

  it("simple", async () => {
    const fpath = path.join(root, "sample.schema.yml");
    fs.writeFileSync(
      fpath,
      `
version: 1
schemas:
- id: pro
  desc: projects
  parent: root
  namespace: true
  children:
    - quickstart
    - concepts
    - tips
    - faq
    - upgrading
    - topic
    - install
- id: quickstart
  desc: getting started with a project
- id: concepts
  desc: basic concepts to do with the project
- id: tips
- id: faq
- id: upgrading
- id: install
- id: topic
  desc: important areas of a project
  namespace: true
`,
      { encoding: "utf-8" }
    );
    const schema = file2Schema(fpath);
    expect(_.values(schema.schemas).length).toEqual(8);
  });
});
