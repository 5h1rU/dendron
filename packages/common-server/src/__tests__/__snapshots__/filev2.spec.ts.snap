// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaModuleProps2File non-root 1`] = `
"version: 1
imports: []
schemas:
  - id: root
    children:
      - ch1
    title: root
    parent: root
  - id: ch1
    children: []
    title: ch1
    parent: root
"
`;

exports[`schemaModuleProps2File root 1`] = `
"version: 1
imports: []
schemas:
  - id: root
    children: []
    title: root
    parent: root
"
`;
